train_gpt2.cu

```diff
@@ -256,6 +260,29 @@ __device__ floatX warpReduceSum(floatX val) {
 }
 #endif
 
+using reduction_func_t = float (*) (float);
+__device__ float blockReduce(float val, reduction_func_t warp_reduction, bool final_sync=false) {
+    // two reductions of up to 1024 threads:
+    // 1) inside warp (shuffle), 2) cross-warp (shared memory), 3) inside warp (shuffle)
+    __shared__ float shared_val[32];
+    int lane_id = threadIdx.x % 32;
+    int warp_id = threadIdx.x / 32;
+    int num_warps = blockDim.x / 32;
+
+    float warp_val = warp_reduction(val);
+    if (lane_id == 0) { shared_val[warp_id] = warp_val; }
+    __syncthreads();
+    // same strategy, now reduce across warps
+    warp_val = (lane_id < num_warps) ? shared_val[lane_id] : 0.0f;
```

Comment:
this assumes that the "base case" of the reduction is 0.0f right? Which may not be the case for arbitrary types of reduction.

Reply:
in light of this comment, we may want to switch from function pointers to functors anyway, like the `cg::plus`  object. Something like
```c++
struct plus {
     static float operator()(float a, float b) { return  a + b; }
     static float neutral() { return 0.f;} 
}
```

Reply:
Agreed, for now I just added an "out_of_bounds" parameter that defaults to 0.0f so it's at least able to handle that case if we need it to.